include/alpaka/mem/view/Accessor.hpp

```diff
@@ -0,0 +1,423 @@
+/* Copyright 2021 Bernhard Manfred Gruber
+
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED “AS IS” AND ISC DISCLAIMS ALL WARRANTIES WITH
+ * REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL ISC BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
+ * IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#pragma once
+
+#include <alpaka/dim/Traits.hpp>
+#include <alpaka/mem/view/Traits.hpp>
+
+#include <tuple>
+
+namespace alpaka
+{
+    namespace internal
+    {
+        template<typename T>
+        auto asBytePtr(T* p)
+        {
+            return reinterpret_cast<char*>(p);
+        }
+        template<typename T>
+        auto asBytePtr(const T* p)
+        {
+            return reinterpret_cast<const char*>(p);
+        }
+    } // namespace internal
+
+    //! Access tag type indicating read-only access.
+    struct ReadAccess
+    {
+    };
+
+    //! Access tag type indicating write-only access.
+    struct WriteAccess
+    {
+    };
+
+    //! Access tag type indicating read-write access.
+    struct ReadWriteAccess
+    {
+    };
+
+    //! An accessor is an abstraction for accessing memory objects such as views and buffers.
+    //! @tparam TMemoryHandle A handle to a memory object.
+    //! @tparam TElem The type of the element stored by the memory object. Values and references to this type are
+    //! returned on access.
+    //! @tparam TBufferIdx The integral type used for indexing and index computations.
+    //! @tparam TDim The dimensionality of the accessed data.
+    //! @tparam TAccessModes Either a single access tag type or a `std::tuple` containing multiple access tag types.
+    template<typename TMemoryHandle, typename TElem, typename TBufferIdx, std::size_t TDim, typename TAccessModes>
```

Comment:
How would I portably declare a kernel parameter list that accepts an accessor with arbitrary `TMemoryHandle`?

```c++
using cuda_accessor = Accessor<float*, float, int, 1, ReadWriteAccess>;
using sycl_accessor = Accessor<sycl::accessor<float, /* other parms */>, float, size_t, 1, ReadWriteAccess>;

struct my_kernel
{
    auto operator()(/* ??? */ accessor_as_parameter)
    {
    }
};
```

Reply:
Just templating the memory handle? Or maybe I did not get your question.
```c++
using cuda_accessor = Accessor<float*, float, int, 1, ReadWriteAccess>;
using sycl_accessor = Accessor<sycl::accessor<float, /* other parms */>, float, size_t, 1, ReadWriteAccess>;

struct my_kernel
{
    template <typename MemoryHandle, typename Idx>
    auto operator()(Accessor<MemoryHandle, float, Idx, 1, ReadWriteAccess> accessor_as_parameter)
    {
    }
};
```

Reply:
Yes, I thought of that. But that introduces an additional template parameter and forces the user to think about our internal memory handles. Wouldn't it be better to specialize the accessor (and its internal handles) through the already existing `TAcc` template parameter?

Reply:
As we discussed today, memory handles are different from accelerators. An accessor backed by a `float*` has the same type, independently if it is used with e.g. CUDA or OpenMP. Furthermore, you can have different accessor types for the same accelerator. E.g. you have two accessors passed to a CUDA kernel, one backed by a CUDA buffer, one backed by a CUDA texture handle. The accessors have different memory handles and thus different types, but run an the same accelerator.

```c++
template <typename Accelerator, typename MemoryHandle1, typename MemoryHandle2, typename Idx>
    auto operator()(const Accelerator& acc, Accessor<MemoryHandle1, float, Idx, 1, ReadWriteAccess> accessor1, , Accessor<MemoryHandle2, float, Idx, 1, ReadWriteAccess> accessor2)
    {
    }
```

In the example above I can switch between textures and buffers as I wish, independently of the accelerator.

Reply:
A few more details: With the current approach, the user code needs to look like this:

```c++
struct user_kernel
{
    template <typename TAcc, typename TMemoryHandle, typename Idx>
    auto operator()(TAcc const& acc, Accessor<TMemoryHandle, float, Idx, ReadWriteAccess>)
    {
    }
};
```

`TAcc` must be present in all sensible alpaka kernels. So it would make sense to use `TAcc` for backend-specific specialization of accessors.

While I'm writing this I noticed that `TMemoryHandle` in user code introduces even more problems. What if I want accelerators for different data types?

```c++
template <typename TAcc, typename TMemoryHandle, typename Idx>
auto operator()(TAcc const& acc,
                Accessor<TMemoryHandle, float, Idx, ReadWriteAccess> floatAcc,
                Accessor<TMemoryHandle, int, Idx, ReadWriteAccess> intAcc // uh-oh
                );
```

Reply:
> In the example above I can switch between textures and buffers as I wish, independently of the accelerator.

SYCL tried something similar in the past, but they ultimately removed this. Now there are distinct accessor types for buffers, local (shared) memory, sampled and unsampled textures. I'm not saying that we have to do the same but having a "one size fits all" accessor might not be the right way to go.

My main issue with this is that the kernel interface from a user point of view is getting blown out of proportion. Take our vectorAdd example:

```c++
// Before
template<typename TAcc, typename TElem, typename TIdx>
ALPAKA_FN_ACC auto operator()(TAcc const& acc,
                              TElem const* const A,
                              TElem const* const B,
                              TElem* const C,
                              TIdx const& numElements) const -> void;

// After
template<typename TAcc, typename TElem, typename TMemoryHandleA, typename TMemoryHandleB, typename TMemoryHandleC, typename TIdx>
ALPAKA_FN_ACC auto operator()(TAcc const& acc,
                              Accessor<TMemoryHandleA, TElem, TIdx, ReadAccess> A,
                              Accessor<TMemoryHandleB, TElem, TIdx, ReadAccess> B,
                              Accessor<TMemoryHandleC, TElem, TIdx, WriteAccess> C,
                              TIdx const& numElements) const -> void;
```

Granted, `TMemoryHandleA` and `TMemoryHandleB` could be merged in this example (as long as none of A's or B's other properties change, so not really safe). But IMO there is still too much verbosity added to user code if one needs to add a separate template parameter for every accessor with a distinct datatype / access mode.

Reply:
> With the current approach, the user code needs to look like this:
> 
> ```c++
> struct user_kernel
> {
>     template <typename TAcc, typename TMemoryHandle, typename Idx>
>     auto operator()(TAcc const& acc, Accessor<TMemoryHandle, float, Idx, ReadWriteAccess>)
>     {
>     }
> };
> ```
> `TAcc` must be present in all sensible alpaka kernels. So it would make sense to use `TAcc` for backend-specific specialization of accessors.

And "this is not even my final form", because you forgot the additional template argument for the dimensionality ;)

We could add the accelerator as a template argument and use the same addressing type (`Idx`) in the accessor, would look like this:
```c++
     template <typename TAcc, typename TMemoryHandle>
     auto operator()(TAcc const& acc, Accessor<TAcc, TMemoryHandle, float, 1, ReadWriteAccess>)
     {
     }
```
That could help marginally.

> While I'm writing this I noticed that `TMemoryHandle` in user code introduces even more problems. What if I want accessors for different data types?

Well of course each accessor needs its own memory handle, so we can have individually different memory handles:

```c++
template <typename TAcc, typename TMemoryHandle1, typename TMemoryHandle2, typename Idx>
auto operator()(TAcc const& acc,
                Accessor<TMemoryHandle1, float, Idx, ReadWriteAccess> floatAcc,
                Accessor<TMemoryHandle2, int, Idx, ReadWriteAccess> intAcc
                );
```
>>In the example above I can switch between textures and buffers as I wish, independently of the accelerator.
> SYCL tried something similar in the past, but they ultimately removed this. Now there are distinct accessor types for buffers, local (shared) memory, sampled and unsampled textures. I'm not saying that we have to do the same but having a "one size fits all" accessor might not be the right way to go.

We had this discussion some while back (I could not find it though) and we decided we would like accessors and textures under the same interface. But I think I also argued weakly against it, because texture usually have different usages. On the other hand, omitting texture support from the accelerator does not simplify its design a lot. We could not get rid of the memory handle, because SYCL still has fancy pointers. So the interface would be the same.

> My main issue with this is that the kernel interface from a user point of view is getting blown out of proportion.

I agree. I would like to neurally restate, as a said many meetings ago, that I still believe that accessors will be a major overcomplication and an impediment for the average user. They are super cool for ND buffers. But ordinary users working on their 1D arrays will suffer with anything more complicated than a `T*`. And we can always wrap a `T*` in an ND array abstraction. I think I mentioned `mdspan<T>` in the past ;)

Reply:
As discussed in today's VC: We may want to think about providing some convenience `using`s to get rid of some template parameters.

Reply:
With a few sensible defaults we could have e.g.:

```c++
template <typename TAcc, typename TElem, size_t Dim, typename AccessModes = ReadWrite>
using BufferAccessor = Accessor<TElem*, TElem, Idx<TAcc>, Dim, AccessModes>; // for SYCL use something else for the memory handle

...

template <typename TAcc>
auto operator()(TAcc const& acc, BufferAccessor<TAcc, float, 1> data, ImageAccessor<TAcc> data)
{
}
```

We could make buffers read/write by default and they are usually backed by a pointer, depending on the backend. Images are typically read-only, 2D and fetch float4. So yes, with some good aliases, I think we could reduce the template parameter list quite a bit.

Reply:
This is something that seems very convenient for both users and us library developers!